File: isegoria/src/main/java/com/eulersbridge/isegoria/friends/FriendAdapter.java
package com.eulersbridge.isegoria.friends;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.network.api.models.Contact;
import com.eulersbridge.isegoria.network.api.models.GenericUser;
import com.eulersbridge.isegoria.network.api.models.Institution;
import com.eulersbridge.isegoria.util.Strings;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class FriendAdapter extends RecyclerView.Adapter<UserViewHolder> implements UserViewHolder.OnClickListener {

    interface Delegate {
        void getContactInstitution(long institutionId, WeakReference<UserViewHolder> weakViewHolder);
        void onContactClick(@NonNull Contact contact);
    }

    final private List<Contact> items = new ArrayList<>();
    final private Delegate delegate;

    FriendAdapter(Delegate delegate) {
        this.delegate = delegate;
    }

    @Override
    public void onViewClick(@Nullable GenericUser user) {
        if (user != null && delegate != null)
            delegate.onContactClick((Contact) user);
    }

    @Override
    public void onActionClick(@Nullable GenericUser user) {
        onViewClick(user);
    }

    @Override
    public UserViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).
                inflate(R.layout.friend_partial_list_item, parent, false);

        return new UserViewHolder(itemView, R.drawable.profileactive, this);
    }

    @Override
    public void onBindViewHolder(UserViewHolder holder, int position) {
        Contact item = items.get(position);
        holder.setItem(item);

        if (delegate != null && item.institutionId != null) {
            WeakReference<UserViewHolder> weakViewHolder = new WeakReference<>(holder);
            delegate.getContactInstitution(item.institutionId, weakViewHolder);
        }
    }

    @Override
    public void onBindViewHolder(UserViewHolder holder, int position, List<Object> payloads) {
        onBindViewHolder(holder, position);
    }

    void setInstitution(@Nullable Institution institution, WeakReference<UserViewHolder> weakViewHolder) {
        UserViewHolder viewHolder = weakViewHolder.get();

        if (viewHolder != null && institution != null)
            viewHolder.setInstitution(institution);
    }

    void setItems(@NonNull List<Contact> newItems) {
        DiffUtil.DiffResult result = DiffUtil.calculateDiff(new DiffUtil.Callback() {
            @Override
            public int getOldListSize() {
                return items.size();
            }

            @Override
            public int getNewListSize() {
                return newItems.size();
            }

            @Override
            public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                Contact oldItem = items.get(oldItemPosition);
                Contact newItem = items.get(newItemPosition);
                return Strings.equal(oldItem.email, newItem.email);
            }

            @Override
            public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                Contact oldItem = items.get(oldItemPosition);
                Contact newItem = items.get(newItemPosition);
                return oldItem.equals(newItem);
            }
        });

        items.clear();
        items.addAll(newItems);

        result.dispatchUpdatesTo(this);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/friends/FriendAdapter.kt
package com.eulersbridge.isegoria.friends

import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.network.api.models.Contact
import com.eulersbridge.isegoria.network.api.models.GenericUser
import com.eulersbridge.isegoria.network.api.models.Institution
import java.lang.ref.WeakReference

class FriendAdapter internal constructor(private val delegate: Delegate?) :
    RecyclerView.Adapter<UserViewHolder>(), UserViewHolder.OnClickListener {

    private val items = mutableListOf<Contact>()

    internal interface Delegate {
        fun getContactInstitution(
            institutionId: Long,
            weakViewHolder: WeakReference<UserViewHolder>
        )

        fun onContactClick(contact: Contact)
    }

    override fun onViewClick(user: GenericUser?) {
        (user as? Contact)?.let {
            delegate?.onContactClick(it)
        }
    }

    override fun onActionClick(user: GenericUser?) = onViewClick(user)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserViewHolder {
        val itemView = LayoutInflater.from(parent.context)
            .inflate(R.layout.friend_partial_list_item, parent, false)

        return UserViewHolder(itemView, R.drawable.profileactive, this)
    }

    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        val item = items[position]
        holder.setItem(item)

        item.institutionId?.let {
            val weakViewHolder = WeakReference(holder)
            delegate?.getContactInstitution(it, weakViewHolder)
        }
    }

    override fun onBindViewHolder(holder: UserViewHolder, position: Int, payloads: MutableList<Any>) {
        onBindViewHolder(holder, position)
    }

    internal fun setInstitution(
        institution: Institution?,
        weakViewHolder: WeakReference<UserViewHolder>
    ) {
        institution?.let {
            weakViewHolder.get()?.setInstitution(it)
        }
    }

    internal fun setItems(newItems: List<Contact>) {
        val result = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize() = items.size

            override fun getNewListSize() = newItems.size

            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                val oldItem = items[oldItemPosition]
                val newItem = items[newItemPosition]
                return oldItem.email == newItem.email
            }

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                val oldItem = items[oldItemPosition]
                val newItem = items[newItemPosition]
                return oldItem == newItem
            }
        })

        items.clear()
        items.addAll(newItems)

        result.dispatchUpdatesTo(this)
    }

    override fun getItemCount() = items.size
}
